File: docs/scripts/pop3-client.py
#!/usr/bin/env python3
"""
POP3 Client Example in Python

This script demonstrates how to use the Python poplib library to interact
with a POP3 server. It shows all major POP3 operations including:
- Connecting to a POP3 server
- Authentication
- Listing messages
- Retrieving messages
- Deleting messages
- Error handling

Usage:
    python3 pop3-client.py

Requirements:
    Python 3.6+
"""

import poplib
from email.parser import Parser
from email.policy import default
import sys


def example_basic_connection():
    """Example 1: Basic connection and authentication"""
    print("\n" + "="*50)
    print("EXAMPLE 1: Basic Connection and Authentication")
    print("="*50 + "\n")
    
    try:
        # Connect to POP3 server
        print("Connecting to localhost:110...")
        pop3_server = poplib.POP3("localhost", 110)
        
        # Read the server greeting
        greeting = pop3_server.getwelcome()
        print(f"Server greeting: {greeting.decode('utf-8')}")
        
        # Authenticate
        print("\nAuthenticating...")
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        print("Authentication successful!")
        
        # Close connection
        pop3_server.quit()
        print("Connection closed")
        
        return True
        
    except poplib.error_proto as e:
        print(f"POP3 protocol error: {e}")
        return False
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_list_messages():
    """Example 2: List all messages in mailbox"""
    print("\n" + "="*50)
    print("EXAMPLE 2: List All Messages")
    print("="*50 + "\n")
    
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        # Get mailbox statistics
        message_count, mailbox_size = pop3_server.stat()
        print(f"Mailbox statistics:")
        print(f"  Messages: {message_count}")
        print(f"  Total size: {mailbox_size} bytes")
        
        # List all messages
        print(f"\nListing all messages:")
        response, message_list, octets = pop3_server.list()
        print(f"Response: {response.decode('utf-8')}")
        
        for msg in message_list:
            msg_info = msg.decode('utf-8')
            parts = msg_info.split()
            if len(parts) >= 2:
                msg_num, msg_size = parts[0], parts[1]
                print(f"  Message {msg_num}: {msg_size} bytes")
        
        pop3_server.quit()
        return True
        
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_retrieve_message():
    """Example 3: Retrieve and parse a message"""
    print("\n" + "="*50)
    print("EXAMPLE 3: Retrieve and Parse Message")
    print("="*50 + "\n")
    
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        message_count, _ = pop3_server.stat()
        
        if message_count > 0:
            print(f"Retrieving message 1...")
            
            # Retrieve message (returns response, lines, octets)
            response, lines, octets = pop3_server.retr(1)
            print(f"Response: {response.decode('utf-8')}")
            print(f"Size: {octets} bytes")
            
            # Parse the message
            msg_content = b'\n'.join(lines)
            msg = Parser(policy=default).parsestr(msg_content.decode('utf-8'))
            
            print(f"\n--- Message Details ---")
            print(f"From: {msg.get('From', 'N/A')}")
            print(f"To: {msg.get('To', 'N/A')}")
            print(f"Subject: {msg.get('Subject', 'N/A')}")
            print(f"Date: {msg.get('Date', 'N/A')}")
            
            # Get message body
            if msg.is_multipart():
                print("\n--- Message Body (multipart) ---")
                for part in msg.walk():
                    if part.get_content_type() == "text/plain":
                        body = part.get_payload(decode=True)
                        if body:
                            print(body.decode('utf-8', errors='ignore')[:500])
                            break
            else:
                print("\n--- Message Body ---")
                body = msg.get_payload(decode=True)
                if body:
                    print(body.decode('utf-8', errors='ignore')[:500])
        else:
            print("No messages in mailbox")
        
        pop3_server.quit()
        return True
        
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_retrieve_all_messages():
    """Example 4: Retrieve all messages"""
    print("\n" + "="*50)
    print("EXAMPLE 4: Retrieve All Messages")
    print("="*50 + "\n")
    
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        message_count, _ = pop3_server.stat()
        print(f"Retrieving {message_count} message(s)...\n")
        
        for i in range(1, message_count + 1):
            print(f"--- Message {i} ---")
            response, lines, octets = pop3_server.retr(i)
            
            msg_content = b'\n'.join(lines)
            msg = Parser(policy=default).parsestr(msg_content.decode('utf-8'))
            
            print(f"Subject: {msg.get('Subject', 'N/A')}")
            print(f"From: {msg.get('From', 'N/A')}")
            print(f"Size: {octets} bytes")
            print()
        
        pop3_server.quit()
        return True
        
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_delete_messages():
    """Example 5: Delete messages"""
    print("\n" + "="*50)
    print("EXAMPLE 5: Delete Messages")
    print("="*50 + "\n")
    
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        message_count, _ = pop3_server.stat()
        
        if message_count > 0:
            print(f"Marking message 1 for deletion...")
            response = pop3_server.dele(1)
            print(f"Response: {response.decode('utf-8')}")
            
            print("\nResetting (unmark deletion)...")
            response = pop3_server.rset()
            print(f"Response: {response.decode('utf-8')}")
            
            # Verify message still exists
            message_count_after_reset, _ = pop3_server.stat()
            print(f"Messages after reset: {message_count_after_reset}")
            
            # Mark for deletion again
            print("\nMarking message 1 for deletion again...")
            response = pop3_server.dele(1)
            print(f"Response: {response.decode('utf-8')}")
            
            print("\nNote: Deletion will be committed when we QUIT")
        else:
            print("No messages to delete")
        
        pop3_server.quit()
        print("Connection closed - deletion committed")
        return True
        
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_top_command():
    """Example 6: Use TOP command to get message headers"""
    print("\n" + "="*50)
    print("EXAMPLE 6: TOP Command (Headers Only)")
    print("="*50 + "\n")
    
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        message_count, _ = pop3_server.stat()
        
        if message_count > 0:
            # Get headers and first 0 lines of body
            print("Retrieving headers of message 1...")
            response, lines, octets = pop3_server.top(1, 0)
            print(f"Response: {response.decode('utf-8')}")
            
            print("\n--- Headers ---")
            for line in lines[:10]:  # Show first 10 lines
                print(line.decode('utf-8', errors='ignore'))
        else:
            print("No messages in mailbox")
        
        pop3_server.quit()
        return True
        
    except poplib.error_proto as e:
        print(f"TOP command not supported by server: {e}")
        return False
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_uidl_command():
    """Example 7: Use UIDL command for unique message IDs"""
    print("\n" + "="*50)
    print("EXAMPLE 7: UIDL Command (Unique IDs)")
    print("="*50 + "\n")
    
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        message_count, _ = pop3_server.stat()
        
        if message_count > 0:
            print("Getting unique IDs for all messages...")
            response, uidl_list, octets = pop3_server.uidl()
            print(f"Response: {response.decode('utf-8')}")
            
            print("\n--- Unique IDs ---")
            for uidl in uidl_list:
                print(uidl.decode('utf-8'))
        else:
            print("No messages in mailbox")
        
        pop3_server.quit()
        return True
        
    except poplib.error_proto as e:
        print(f"UIDL command not supported by server: {e}")
        return False
    except Exception as e:
        print(f"Error: {e}")
        return False


def example_error_handling():
    """Example 8: Error handling scenarios"""
    print("\n" + "="*50)
    print("EXAMPLE 8: Error Handling")
    print("="*50 + "\n")
    
    # Test 1: Wrong credentials
    print("--- Test 1: Authentication with wrong credentials ---")
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("wronguser")
        pop3_server.pass_("wrongpass")
        pop3_server.quit()
        print("Unexpected: Authentication succeeded")
    except poplib.error_proto as e:
        print(f"Expected error: {e}")
    except Exception as e:
        print(f"Connection error (server not running?): {e}")
    
    # Test 2: Retrieve non-existent message
    print("\n--- Test 2: Retrieve non-existent message ---")
    try:
        pop3_server = poplib.POP3("localhost", 110)
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        
        # Try to retrieve message 99999 (should not exist)
        response, lines, octets = pop3_server.retr(99999)
        print("Unexpected: Retrieved non-existent message")
        pop3_server.quit()
    except poplib.error_proto as e:
        print(f"Expected error: {e}")
    except Exception as e:
        print(f"Connection error: {e}")
    
    # Test 3: Connection to wrong port
    print("\n--- Test 3: Connection to wrong port ---")
    try:
        pop3_server = poplib.POP3("localhost", 9999, timeout=2)
        print("Unexpected: Connection succeeded")
    except Exception as e:
        print(f"Expected error: {e}")


def example_secure_connection():
    """Example 9: Secure POP3 connection (POP3S)"""
    print("\n" + "="*50)
    print("EXAMPLE 9: Secure POP3 Connection (POP3S)")
    print("="*50 + "\n")
    
    try:
        # POP3S typically uses port 995
        print("Connecting to localhost:995 with SSL/TLS...")
        pop3_server = poplib.POP3_SSL("localhost", 995)
        
        greeting = pop3_server.getwelcome()
        print(f"Server greeting: {greeting.decode('utf-8')}")
        
        pop3_server.user("testuser")
        pop3_server.pass_("testpass")
        print("Secure authentication successful!")
        
        pop3_server.quit()
        return True
        
    except Exception as e:
        print(f"POP3S connection failed (expected if not configured): {e}")
        return False


def main():
    """Run all examples"""
    print("="*60)
    print("POP3 Client Examples in Python")
    print("="*60)
    print("\nThese examples demonstrate POP3 operations using Python's")
    print("poplib library.")
    print("\nNote: Examples assume a POP3 server is running on localhost:110")
    print("      with user 'testuser' and password 'testpass'.")
    print("\nIf the server is not running, you'll see connection errors.")
    
    # Run examples
    examples = [
        example_basic_connection,
        example_list_messages,
        example_retrieve_message,
        example_retrieve_all_messages,
        example_delete_messages,
        example_top_command,
        example_uidl_command,
        example_error_handling,
        example_secure_connection,
    ]
    
    for example_func in examples:
        try:
            example_func()
        except KeyboardInterrupt:
            print("\n\nInterrupted by user")
            sys.exit(0)
        except Exception as e:
            print(f"Unexpected error in {example_func.__name__}: {e}")
    
    print("\n" + "="*60)
    print("All examples completed!")
    print("="*60)


if __name__ == "__main__":
    main()

File: examples/pop3_client.rs
//! POP3 Client Example
//!
//! This example demonstrates how to connect to a POP3 server and perform
//! common operations such as authentication, listing messages, retrieving
//! messages, and deleting messages.
//!
//! Usage:
//!   cargo run --example pop3_client
//!
//! This example shows:
//! - Connecting to a POP3 server
//! - Authenticating with USER and PASS commands
//! - Getting mailbox statistics with STAT
//! - Listing messages with LIST
//! - Retrieving messages with RETR
//! - Deleting messages with DELE
//! - Properly closing connection with QUIT

use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;

/// Represents a POP3 client connection
struct Pop3Client {
    stream: TcpStream,
    reader: BufReader<TcpStream>,
}

impl Pop3Client {
    /// Connect to a POP3 server
    fn connect(host: &str, port: u16) -> Result<Self, Box<dyn std::error::Error>> {
        println!("Connecting to {}:{}...", host, port);
        let stream = TcpStream::connect(format!("{}:{}", host, port))?;
        let reader = BufReader::new(stream.try_clone()?);
        
        let mut client = Pop3Client { stream, reader };
        
        // Read the server greeting
        let greeting = client.read_response()?;
        println!("Server greeting: {}", greeting);
        
        if !greeting.starts_with("+OK") {
            return Err(format!("Server did not send +OK greeting: {}", greeting).into());
        }
        
        Ok(client)
    }
    
    /// Send a command to the server
    fn send_command(&mut self, command: &str) -> Result<(), Box<dyn std::error::Error>> {
        println!("> {}", command);
        writeln!(self.stream, "{}", command)?;
        self.stream.flush()?;
        Ok(())
    }
    
    /// Read a single-line response from the server
    fn read_response(&mut self) -> Result<String, Box<dyn std::error::Error>> {
        let mut response = String::new();
        self.reader.read_line(&mut response)?;
        let response = response.trim_end().to_string();
        println!("< {}", response);
        Ok(response)
    }
    
    /// Read a multi-line response from the server (ends with ".")
    fn read_multiline_response(&mut self) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        let mut lines = Vec::new();
        loop {
            let mut line = String::new();
            self.reader.read_line(&mut line)?;
            let line = line.trim_end().to_string();
            println!("< {}", line);
            
            if line == "." {
                break;
            }
            lines.push(line);
        }
        Ok(lines)
    }
    
    /// Authenticate with username and password
    fn authenticate(&mut self, username: &str, password: &str) -> Result<(), Box<dyn std::error::Error>> {
        println!("\n=== Authenticating ===");
        
        // Send USER command
        self.send_command(&format!("USER {}", username))?;
        let response = self.read_response()?;
        if !response.starts_with("+OK") {
            return Err(format!("USER command failed: {}", response).into());
        }
        
        // Send PASS command
        self.send_command(&format!("PASS {}", password))?;
        let response = self.read_response()?;
        if !response.starts_with("+OK") {
            return Err(format!("PASS command failed: {}", response).into());
        }
        
        println!("Authentication successful!");
        Ok(())
    }
    
    /// Get mailbox statistics (STAT command)
    fn stat(&mut self) -> Result<(usize, usize), Box<dyn std::error::Error>> {
        println!("\n=== Getting mailbox statistics ===");
        
        self.send_command("STAT")?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("STAT command failed: {}", response).into());
        }
        
        // Parse response: +OK <message_count> <mailbox_size>
        let parts: Vec<&str> = response.split_whitespace().collect();
        if parts.len() >= 3 {
            let message_count = parts[1].parse::<usize>()?;
            let mailbox_size = parts[2].parse::<usize>()?;
            println!("Mailbox has {} messages ({} bytes)", message_count, mailbox_size);
            Ok((message_count, mailbox_size))
        } else {
            Err("Invalid STAT response format".into())
        }
    }
    
    /// List all messages (LIST command)
    fn list(&mut self) -> Result<Vec<(usize, usize)>, Box<dyn std::error::Error>> {
        println!("\n=== Listing all messages ===");
        
        self.send_command("LIST")?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("LIST command failed: {}", response).into());
        }
        
        let lines = self.read_multiline_response()?;
        let mut messages = Vec::new();
        
        for line in lines {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 {
                let msg_num = parts[0].parse::<usize>()?;
                let msg_size = parts[1].parse::<usize>()?;
                messages.push((msg_num, msg_size));
                println!("  Message {}: {} bytes", msg_num, msg_size);
            }
        }
        
        Ok(messages)
    }
    
    /// List a specific message (LIST command with message number)
    fn list_message(&mut self, msg_num: usize) -> Result<usize, Box<dyn std::error::Error>> {
        println!("\n=== Listing message {} ===", msg_num);
        
        self.send_command(&format!("LIST {}", msg_num))?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("LIST command failed: {}", response).into());
        }
        
        // Parse response: +OK <msg_num> <msg_size>
        let parts: Vec<&str> = response.split_whitespace().collect();
        if parts.len() >= 3 {
            let msg_size = parts[2].parse::<usize>()?;
            println!("Message {} size: {} bytes", msg_num, msg_size);
            Ok(msg_size)
        } else {
            Err("Invalid LIST response format".into())
        }
    }
    
    /// Retrieve a message (RETR command)
    fn retrieve(&mut self, msg_num: usize) -> Result<String, Box<dyn std::error::Error>> {
        println!("\n=== Retrieving message {} ===", msg_num);
        
        self.send_command(&format!("RETR {}", msg_num))?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("RETR command failed: {}", response).into());
        }
        
        let lines = self.read_multiline_response()?;
        let message = lines.join("\n");
        println!("Retrieved message ({} lines)", lines.len());
        
        Ok(message)
    }
    
    /// Delete a message (DELE command)
    fn delete(&mut self, msg_num: usize) -> Result<(), Box<dyn std::error::Error>> {
        println!("\n=== Deleting message {} ===", msg_num);
        
        self.send_command(&format!("DELE {}", msg_num))?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("DELE command failed: {}", response).into());
        }
        
        println!("Message {} marked for deletion", msg_num);
        Ok(())
    }
    
    /// Reset the session (RSET command) - unmarks deleted messages
    fn reset(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        println!("\n=== Resetting session ===");
        
        self.send_command("RSET")?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("RSET command failed: {}", response).into());
        }
        
        println!("Session reset - all delete marks removed");
        Ok(())
    }
    
    /// No-op command (NOOP) - keeps connection alive
    fn noop(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        println!("\n=== Sending NOOP ===");
        
        self.send_command("NOOP")?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("NOOP command failed: {}", response).into());
        }
        
        println!("NOOP successful");
        Ok(())
    }
    
    /// Close the connection (QUIT command)
    fn quit(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        println!("\n=== Closing connection ===");
        
        self.send_command("QUIT")?;
        let response = self.read_response()?;
        
        if !response.starts_with("+OK") {
            return Err(format!("QUIT command failed: {}", response).into());
        }
        
        println!("Connection closed successfully");
        Ok(())
    }
}

/// Example 1: Complete email retrieval session
fn example_complete_session() -> Result<(), Box<dyn std::error::Error>> {
    println!("\n========================================");
    println!("EXAMPLE 1: Complete Email Retrieval Session");
    println!("========================================\n");
    
    let mut client = Pop3Client::connect("localhost", 110)?;
    
    // Authenticate
    client.authenticate("testuser", "testpass")?;
    
    // Get mailbox statistics
    let (message_count, _mailbox_size) = client.stat()?;
    
    // List all messages
    let _messages = client.list()?;
    
    // Retrieve first message if available
    if message_count > 0 {
        let message_content = client.retrieve(1)?;
        println!("\nFirst message preview:");
        println!("---");
        for (i, line) in message_content.lines().take(10).enumerate() {
            println!("{}", line);
            if i == 9 && message_content.lines().count() > 10 {
                println!("... (truncated)");
            }
        }
        println!("---");
    }
    
    // Close connection
    client.quit()?;
    
    Ok(())
}

/// Example 2: Handling multiple messages
fn example_multiple_messages() -> Result<(), Box<dyn std::error::Error>> {
    println!("\n========================================");
    println!("EXAMPLE 2: Handling Multiple Messages");
    println!("========================================\n");
    
    let mut client = Pop3Client::connect("localhost", 110)?;
    client.authenticate("testuser", "testpass")?;
    
    // Get all messages
    let messages = client.list()?;
    
    println!("\nProcessing {} messages:", messages.len());
    for (msg_num, msg_size) in messages {
        println!("\nProcessing message {} ({} bytes):", msg_num, msg_size);
        
        // Retrieve each message
        match client.retrieve(msg_num) {
            Ok(content) => {
                println!("  Successfully retrieved {} bytes", content.len());
                // Process message content here
            }
            Err(e) => {
                println!("  Error retrieving message: {}", e);
            }
        }
    }
    
    client.quit()?;
    Ok(())
}

/// Example 3: Message deletion workflow
fn example_deletion_workflow() -> Result<(), Box<dyn std::error::Error>> {
    println!("\n========================================");
    println!("EXAMPLE 3: Message Deletion Workflow");
    println!("========================================\n");
    
    let mut client = Pop3Client::connect("localhost", 110)?;
    client.authenticate("testuser", "testpass")?;
    
    let (message_count, _) = client.stat()?;
    
    if message_count > 0 {
        // Mark first message for deletion
        client.delete(1)?;
        
        // Change our mind - reset to unmark
        client.reset()?;
        
        // Verify message is still there
        let messages = client.list()?;
        println!("\nAfter reset, mailbox still has {} messages", messages.len());
        
        // Delete again (this time for real)
        client.delete(1)?;
        
        // Quit to commit the deletion
        println!("\nNote: Deletion will be committed when we QUIT");
    } else {
        println!("No messages to delete");
    }
    
    client.quit()?;
    Ok(())
}

/// Example 4: Error handling
fn example_error_handling() -> Result<(), Box<dyn std::error::Error>> {
    println!("\n========================================");
    println!("EXAMPLE 4: Error Handling");
    println!("========================================\n");
    
    // Example: Authentication failure
    println!("--- Testing authentication failure ---");
    match Pop3Client::connect("localhost", 110) {
        Ok(mut client) => {
            match client.authenticate("wronguser", "wrongpass") {
                Ok(_) => println!("Unexpected: Authentication succeeded"),
                Err(e) => println!("Expected error: {}", e),
            }
            let _ = client.quit();
        }
        Err(e) => {
            println!("Connection failed (expected if server not running): {}", e);
        }
    }
    
    // Example: Invalid message number
    println!("\n--- Testing invalid message number ---");
    match Pop3Client::connect("localhost", 110) {
        Ok(mut client) => {
            if client.authenticate("testuser", "testpass").is_ok() {
                match client.retrieve(99999) {
                    Ok(_) => println!("Unexpected: Retrieved non-existent message"),
                    Err(e) => println!("Expected error: {}", e),
                }
                let _ = client.quit();
            }
        }
        Err(e) => {
            println!("Connection failed (expected if server not running): {}", e);
        }
    }
    
    Ok(())
}

/// Example 5: Connection error scenarios
fn example_connection_errors() {
    println!("\n========================================");
    println!("EXAMPLE 5: Connection Error Scenarios");
    println!("========================================\n");
    
    // Try to connect to non-existent server
    println!("--- Testing connection to non-existent server ---");
    match Pop3Client::connect("localhost", 9999) {
        Ok(_) => println!("Unexpected: Connection succeeded"),
        Err(e) => println!("Expected error: {}", e),
    }
    
    // Try to connect with invalid hostname
    println!("\n--- Testing connection to invalid hostname ---");
    match Pop3Client::connect("invalid.example.nonexistent", 110) {
        Ok(_) => println!("Unexpected: Connection succeeded"),
        Err(e) => println!("Expected error: {}", e),
    }
}

fn main() {
    println!("===========================================");
    println!("POP3 Client Examples");
    println!("===========================================");
    println!("\nThis example demonstrates various POP3 operations.");
    println!("Note: These examples assume a POP3 server is running on localhost:110");
    println!("      with user 'testuser' and password 'testpass'.");
    println!("\nIf the server is not running, you'll see connection errors,");
    println!("which is normal and demonstrates error handling.\n");
    
    // Run all examples
    // Note: In real usage, you'd typically only run one session at a time
    
    if let Err(e) = example_complete_session() {
        println!("Example 1 error: {}", e);
    }
    
    if let Err(e) = example_multiple_messages() {
        println!("Example 2 error: {}", e);
    }
    
    if let Err(e) = example_deletion_workflow() {
        println!("Example 3 error: {}", e);
    }
    
    if let Err(e) = example_error_handling() {
        println!("Example 4 error: {}", e);
    }
    
    example_connection_errors();
    
    println!("\n===========================================");
    println!("Examples completed!");
    println!("===========================================");
}
